### ISSUE TITLE
[BUG] Rebooting a node makes VMs unable to start

### ISSUE THREAD

**Issue by gregoryg** [harvester/harvester#666]
**Describe the bug**
After rebooting a node, VMs will not start.  New VMs will not start either.

**To Reproduce**
    1.  Install from .iso
        -   Create a new Harvester cluster
        -   Choose installation target: sda 953.9G
        -   hostname: protomolecule,NIC eth0; static 172.16.17.16/24; 172.16.17.1; 172.16.17.5,8.8.8.8
        -   cluster token: protomolecule
        -   password, no SSH key import, no proxy, no Harvester config
    2.  Upload a bootable `qcow2` image Virtual Machines -> Create
    3.  Create multiple-instance VM
        -   Select Multiple Instance
        -   Prefix Name: leap-cluster
        -   count: 6
        -   Do **not** check "Use VM Template"
        -   CPU: 2; Memory: 2
        -   Image: opensuse-leap-15.2.x86\_64-nocloud.qcow2
        -   SSHKey: k3s-local
        -   Volumes: change size from 10 to 40GiB
        -   Networks: no changes (default/virtio/masquerade)
        -   Advance options: add the following cloud config

 ```yaml
#cloud-config
groups:
  - sudo
  - wheel
ssh_authorized_keys:
  - >
    ssh-rsa
    AAAAB3NzaC1/blah/blah
    iamk8s-local
users:
  - default
  - name: gregj
    groups: users, sudo, wheel
    ssh_authorized_keys:
      - >
        ssh-rsa
            AAAAB3NzaC1/blah/blah
 ```
        -   All 6 VMs running in a few minutes - no problems
            -   Events from one of the VMs

                Events

                | Reason                 | Resource                               | Message                                                                                                                                                                                                      | Date         |
                |---------------------- |-------------------------------------- |------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |------------ |
                | Created                | VirtualMachineInstance leap-cluster-1  | VirtualMachineInstance defined.                                                                                                                                                                              | 2.4 mins ago |
                | Started                | VirtualMachineInstance leap-cluster-1  | VirtualMachineInstance started.                                                                                                                                                                              | 3.7 mins ago |
                | SuccessfulMountVolume  | Pod virt-launcher-leap-cluster-1-fnw2k | MapVolume.MapPodDevice succeeded for volume "pvc-66a6a2df-0cfa-493a-b24a-16906c176942" globalMapPath "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-66a6a2df-0cfa-493a-b24a-16906c176942/dev" | 3.7 mins ago |
                | SuccessfulMountVolume  | Pod virt-launcher-leap-cluster-1-fnw2k | MapVolume.MapPodDevice succeeded for volume "pvc-66a6a2df-0cfa-493a-b24a-16906c176942" volumeMapPath "/var/lib/kubelet/pods/c7ed02a2-1d5c-4e6e-978d-8495c33c1caf/volumeDevices/kubernetes.io~csi"            | 3.7 mins ago |
                | Pulled                 | Pod virt-launcher-leap-cluster-1-fnw2k | Container image "quay.io/kubevirt/virt-launcher:v0.38.1" already present on machine                                                                                                                          | 3.7 mins ago |
                | Created                | Pod virt-launcher-leap-cluster-1-fnw2k | Created container compute                                                                                                                                                                                    | 3.7 mins ago |
                | Started                | Pod virt-launcher-leap-cluster-1-fnw2k | Started container compute                                                                                                                                                                                    | 3.7 mins ago |
                | SuccessfulAttachVolume | Pod virt-launcher-leap-cluster-1-fnw2k | AttachVolume.Attach succeeded for volume "pvc-66a6a2df-0cfa-493a-b24a-16906c176942"                                                                                                                          | 3.7 mins ago |
                | SuccessfulCreate       | VirtualMachineInstance leap-cluster-1  | Created virtual machine pod virt-launcher-leap-cluster-1-fnw2k                                                                                                                                               | 4.5 mins ago |
                | SuccessfulCreate       | VirtualMachine leap-cluster-1          | Started the virtual machine by creating the new virtual machine instance leap-cluster-1                                                                                                                      | 4.5 mins ago |
        -   Longhorn in good shape after setting replica count on volumes to 1
   4. NOW: reboot with no other prep

```bash
ssh rancher@protomolecule sudo reboot
```
   5.   Refresh Harvester UI once rebooted
        -   NO VMs start - looking at UIs after 60 minutes
            -   the **only** Longhorn volume in a healthy state is the one used for Minio ![img](./img/01-longhorn-post-reboot.png)

![image](https://user-images.githubusercontent.com/162285/114108514-74f72a80-9890-11eb-9afe-34e99ecbb4e7.png)

![image](https://user-images.githubusercontent.com/162285/114108544-83454680-9890-11eb-90e3-382a6d37e7df.png)

![image](https://user-images.githubusercontent.com/162285/114108556-893b2780-9890-11eb-8dc8-59a0ff8ffcde.png)

   7. Restart one VM from UI to see if it can successfully start
            -   After another 30 minutes - nope it cannot restart

   8. Start a new VM single instance using the same image as above
            -   5 minutes later, VM has not started; events show FailedAttachVolume and FailedMount errors
   9.  stop then delete ALL VMs along with their volumes
            -   5 minutes later, VM still "starting" with attach volume errors
   10. Collect logs
            -   k3s

```bash
wget -O- https://raw.githubusercontent.com/rancherlabs/support-tools/master/collection/rancher/v2.x/logs-collector/rancher2_logs_collector.sh | sudo bash -s
```
            -   Longhorn
```bash
# specify NodePort to service longhorn-backend
HOSTPORT=`hostname`:30487
# initiate bundle generation
bundleinfo=$(curl -s -X POST \
  -H 'Content-Type: application/json' -d '{ "issueURL": "https://github.com/longhorn/longhorn/issues/2118", "description": "2118" }' \
  http://${HOSTPORT}/v1/supportbundles)
bundleid=$(echo $bundleinfo | jq -r '.id')
bundlename=$(echo $bundleinfo | jq -r '.name')

# check download readiness
bundlestate=$(curl -s  -X GET \
   http://${HOSTPORT}/v1/supportbundles/${bundleid}/${bundlename} | jq -r '.state')
# When ready, download the beast
if [ ${bundlestate} == 'ReadyForDownload' ] ; then
    curl -s -X GET \
         http://${HOSTPORT}/v1/supportbundles/${bundleid}/${bundlename}/download --output /tmp/support-bundle.zip
    echo "Wrote Longhorn support bundle to /tmp/support-bundle.zip"
fi
```

**Expected behavior**
Longhorn should be able to restart and attach the volumes after an unplanned outage.  Harvester VMs should restart following an unplanned outage.   One note: I have observed this problem in a multi-node Harvester cluster following the reboot of *any* node.   The result was that new VMs were not able to start.

**Logs**
[support-bundle.zip](https://github.com/rancher/harvester/files/6282279/support-bundle.zip)
[protomolecule-2021-04-08_22_49_15.tar.gz](https://github.com/rancher/harvester/files/6282280/protomolecule-2021-04-08_22_49_15.tar.gz)

**Environment:**
 - Harvester ISO version: 0.2.0-rc1
 - Installation Mode: ISO mode
 - Underlying Infrastructure: Bare metal (Intel NUC)

**Comment by guangbochen** [comment]
might be related to https://github.com/rancher/harvester/issues/543

**Comment by gitlawr** [comment]
> Longhorn in good shape after setting replica count on volumes to 1

How did you change the replica count? We generate a storage class for each image. If you change the replica count in the default longhorn storage class it may be irrelevant to this issue

**Comment by gitlawr** [comment]
I can reproduce in v0.2.0-rc1.
Simplified steps:
1. Install single node Harvester by ISO image.
2. Create a VM image using https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-amd64.img
3. Create a VM from the image. Wait until it is running.
4. reboot the node.

**Result**
The VM stays in creating status.

Log from longhorn manager:
```
time="2021-04-13T03:33:42Z" level=debug msg="Replica pvc-4ca71e7b-6f05-4e82-b273-cc1f999e5550-r-ed7e5b58 is waiting for backing image image-kbx8h downloading file to node mydevhost disk c87fc16a-7038-4452-9710-49f7a3a22690 from URL http://minio.harvester-system:9000/vm-images/image-kbx8h, the current state is downloading"
```
The backing-image pod never turns to ready status:
```
$ kubectl -n longhorn-system get po -l longhorn.io/backing-image
NAME                            READY   STATUS    RESTARTS   AGE
image-kbx8h-download-c87fc16a   0/1     Running   0          17m
$ kubectl -n longhorn-system logs -f image-kbx8h-download-c87fc16a
rm: cannot remove '/backing-image/backing': No such file or directory
rm: cannot remove '/backing-image/downloaded': No such file or directory

```
Workaround in v0.2.0-rc1: 
```
$ kubectl -n longhorn-system delete po -l longhorn.io/backing-image
```
The pods will restart and be ready, then the VMs will be running again.

**Comment by gitlawr** [comment]
Validated that this is fixed in master-95a5828a-head(0413), with minio removal and backingimage implementation change.

Following the above steps VMs recover as expected.

Also tested the case when image hosting service is down(like minio service down during reboot in v0.2.0-rc1).
1. Serve the image file in HTTP service.(e.g. cd to the directory that keeps the file and run `docker run -p 80:80 -d -v $(pwd):/usr/share/nginx/html/data lawr/serve`, then the image is available via http://your-server-ip/bionic-server-cloudimg-amd64.img)
2. Install single node Harvester by ISO image.
3. Create a VM image using http://your-server-ip/bionic-server-cloudimg-amd64.img
4. Create a VM from the image. Wait until it is running.
5. shutdown the image service.
4. reboot the node.

**Result**
The VMs are running after the reboot.

**Comment by gregoryg** [comment]
> > Longhorn in good shape after setting replica count on volumes to 1
> 
> How did you change the replica count? We generate a storage class for each image. If you change the replica count in the default longhorn storage class it may be irrelevant to this issue

If I have fewer than 3 nodes, I change each volume's replica count down to the number of nodes so that Longhorn does not report the volume as unhealthy.

**Comment by gregoryg** [comment]
looks like this is fixed in 0.2.0-rc2!

**Comment by TangStone** [comment]
**Verify the following scenarios in version master-0008ce84-head**

- [x] 1. restart one of the nodes and check the status of the vm
- [x] 2. restart all nodes and check the status of the vm
- [x] 3. after reboot, create a vm, check the status of the vm
- [x] 4. shut down the image host service, restart the node, and check the status of the vm
